### PR TITLE
Sc docs

### PR DESCRIPTION
**Pull request by peterjah**

### CODE REVIEW

File: docs/build/smart-contract/basic-concepts/constructor.mdx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    The Args utility can be used to handle serialization and deserialization of these parameters, ensuring consistency with other functions in the contract.
+
+
+###  Example Constructor Implementation
+Below is an example of a constructor function that takes a string parameter for initialization. This example demonstrates how to set up the constructor to accept parameters, verify that it’s only called during deployment, and store the initialized data:
+
+```ts
+import { Args, Storage, Context } from "@massalabs/as-types";
+
+export function constructor(ArgsData: StaticArray<u8>): void {
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/constructor.mdx:23]:
```suggestion
export function constructor(argsData: StaticArray<u8>): void {
```

File: docs/build/smart-contract/basic-concepts/constructor.mdx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    // Ensure the function is only called during deployment
+    assert(Context.isDeployingContract(), "Constructor can only be called during deployment");
+
+    // Deserialize the initialization parameter
+    const initialMessage = new Args(ArgsData).nextString().expect("Initialization parameter missing or invalid");
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/constructor.mdx:28]:
```suggestion
    const initialMessage = new Args(argsData).nextString().expect("Initialization parameter missing or invalid");
```

File: docs/build/smart-contract/basic-concepts/constructor.mdx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+1. **Deployment-Only Execution**:
+    - To ensure that the constructor function is only executed during deployment, include the line `assert(Context.isDeployingContract());` at the start of the function. This assertion checks that the contract is currently in the deployment phase and will throw an error if the function is called after deployment.
+
+2. **Passing Parameters to the Constructor**:
+    - Just like other functions, parameters can be passed to the constructor using serialized data. This allows for the flexible initialization of the contract, enabling the deployment process to be customized with specific data.
+    The Args utility can be used to handle serialization and deserialization of these parameters, ensuring consistency with other functions in the contract.
+
+
+###  Example Constructor Implementation
+Below is an example of a constructor function that takes a string parameter for initialization. This example demonstrates how to set up the constructor to accept parameters, verify that it’s only called during deployment, and store the initialized data:
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/constructor.mdx:18]:
```suggestion
Below is an example of a constructor function that takes a string parameter for initialization. This example demonstrates how to set up the constructor to accept parameters, ensuring it is only called during deployment, and store the initialized data:
```

File: docs/build/smart-contract/basic-concepts/constructor.mdx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+- An Account instance is created using Account.fromEnv(), which retrieves the necessary credentials from environment variables. This ensures the deployment uses a valid account on the Massa network.
+
+3. **Provider Initialization**:
+
+- Web3Provider.buildnet(account) sets up a provider instance connected to the Massa buildnet, allowing the account to interact with the network.
+
+4. **Bytecode Retrieval**:
+
+- The compiled smart contract bytecode is loaded using getScByteCode('build', 'main.wasm'). This function reads the compiled .wasm file from the specified directory.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/constructor.mdx:79]: specify what is 'build' in getScByteCode

File: docs/build/smart-contract/basic-concepts/data-types.mdx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+sidebar_position: 2
+---
+
+# Smart Contract Data Types
+
+When developing smart contracts on Massa, it’s essential to understand the data types available in AssemblyScript, as they directly influence how you handle and manipulate data within the contract. AssemblyScript offers a range of data types that align closely with WebAssembly's capabilities, providing efficient and predictable performance.
+
+## AssemblyScript native types
+
+- i8 / u8: 8-bit signed/unsigned integers
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/data-types.mdx:11]: do a table

File: docs/build/smart-contract/basic-concepts/data-types.mdx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+---
+sidebar_position: 2
+---
+
+# Smart Contract Data Types
+
+When developing smart contracts on Massa, it’s essential to understand the data types available in AssemblyScript, as they directly influence how you handle and manipulate data within the contract. AssemblyScript offers a range of data types that align closely with WebAssembly's capabilities, providing efficient and predictable performance.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/data-types.mdx:7]:
```suggestion
When developing smart contracts on Massa, it is essential to understand the data types available in AssemblyScript, as they directly influence how you handle and manipulate data within the contract. AssemblyScript offers a range of data types that align closely with WebAssembly's capabilities, providing efficient and predictable performance.
```

File: docs/build/smart-contract/basic-concepts/data-types.mdx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+## AssemblyScript native types
+
+- i8 / u8: 8-bit signed/unsigned integers
+- i16 / u16: 16-bit signed/unsigned integers
+- i32 / u32: 32-bit signed/unsigned integers.
+- i64 / u64: 64-bit signed/unsigned integers.
+- f32 / f64: 32-bit and 64-bit floating-point numbers.
+- bool: Represents a boolean value (true or false).
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/data-types.mdx:16]:
```suggestion
- bool: Represents a Boolean value (true or false).
```

File: docs/build/smart-contract/basic-concepts/data-types.mdx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+- string: Represents UTF-16 encoded text.
+
+### Array:
+`Array<T>`: Represents a dynamically-sized collection of elements of type T.
+Arrays are suitable for storing lists of data, such as user addresses or transaction histories. You can define arrays with specific types, such as `Array<i32>` or `Array<string>`, to ensure data consistency.
+
+### Byte array:
+Massa use assemblyScript `StaticArray<u8>` native type to represent data bytes. This type is used internally in storage and is also used to serialize public function parameters and returned data.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/data-types.mdx:24]:
```suggestion
Massa use AssemblyScript `StaticArray<u8>` native type to represent data bytes. This type is used internally in storage and is also used to serialize public function parameters and returned data.
```

File: docs/build/smart-contract/basic-concepts/data-types.mdx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+## Handling Large Numbers in Massa Smart Contracts
+
+In some smart contracts, you may need to work with numbers that exceed the size limitations of standard 64-bit integers. For example, financial applications involving token balances or computations with very large numbers require higher precision and larger ranges than what i64 or u64 can provide. To handle such cases, Massa smart contracts can use the [as-bignum](https://github.com/MaxGraey/as-bignum) library, which supports fixed length big number arithmetic.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/data-types.mdx:28]:
```suggestion
In some smart contracts, you may need to work with numbers that exceed the size limitations of standard 64-bit integers. For example, financial applications involving token balances or computations with very large numbers require higher precision and larger ranges than i64 or u64 can't provide. To handle such cases, Massa smart contracts can use the [as-bignum](https://github.com/MaxGraey/as-bignum) library, which supports fixed length big number arithmetic.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  console.log('Event message:', event.data);
+}
+```
+
+## Retrieving Events with filters
+When retrieving events from a Massa smart contract, you can apply various filters to specify which events to fetch. These filters allow you to narrow down the results based on specific criteria, such as the contract address, caller address, event time range, and whether the events are final or error-related. This flexibility helps you target the most relevant events for monitoring, debugging, or analysis.
+
+### Available Event Filters
+
+Here are the different filters you can use when retrieving events:
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:61]: further specify that each is optional

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 11]
```diff
@@ -0,0 +1,140 @@
+---
+sidebar_position: 8
+---
+
+# Events
+
+In Massa smart contracts, events provide a way to emit runtime messages that can be useful for monitoring, debugging, and logging activities within the contract. Events can be triggered by errors, such as assertion failures or insufficient coin balances, as well as by custom, user-defined events.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:7]:
```suggestion
In Massa's smart contracts, events provide a way to emit runtime messages that can be useful for monitoring, debugging, and logging activities within the contract. Events can be triggered by errors, such as assertion failures or insufficient coin balances, as well as custom, user-defined events.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+## Types of Events
+
+1. **Error Events**:
+
+    - Error events occur when the contract encounters runtime issues, such as:
+      - Failed assertions, where the assert statement fails.
+      - Errors in argument deserialization
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:15]:
```suggestion
      - Errors during argument deserialization
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      - Insufficient coins to write to storage or to call another contract that requires a fee.
+    - These events help identify issues and diagnose why a transaction or contract call may have failed.
+
+2. **User-Defined Events**:
+
+    - In addition to system-generated error events, developers can define custom events to log specific actions or important milestones within the contract.
+    - User-defined events are typically used to provide visibility into contract behavior, such as when a transfer occurs or when a token is minted.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:22]:
```suggestion
    - User-defined events are typically used to provide visibility into contract behavior, for example during a transfer or token minting.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+## Important Considerations for Events
+In Massa, events are not stored on the blockchain ledger, meaning they are ephemeral. Therefore, events should not be used as a source of persistent data. Instead, they are suitable for temporary tasks such as monitoring, logging, or providing real-time feedback.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:25]:
```suggestion
In Massa, events are not stored in the blockchain ledger, meaning they are ephemeral. Therefore, events should not be used as a source of persistent data. Instead, they are suitable for temporary tasks such as monitoring, logging, or providing real-time feedback.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+For usecases that need data persistency, storage should be used.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:26]:
```suggestion
For use cases that require persistent data, storage should be used.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+import { generateEvent } from "@massalabs/massa-as-sdk";
+
+export function transferTokens(to: string, amount: u64): void {
+    // Transfer logic here
+
+    // Emit a custom event after transfer
+    generateEvent(`Transferred ${amount.toString()} tokens to ${to}`);
+}
+```
+In this example, the generateEvent function is used to emit a custom message after a token transfer operation. This message can then be retrieved as shown in the retrieval example below.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:41]:
```suggestion
In this example, the generateEvent function is used to emit a custom message after a token transfer operation. This message can be retrieved as shown in the example below.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+## Retrieving Events
+Although events are not permanently recorded on the blockchain, they can still be retrieved for monitoring purposes after being emitted. Here’s an example of how to retrieve events for a specific smart contract using the [@massalabs/massa-web3](https://github.com/massalabs/massa-web3) provider:
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:44]:
```suggestion
Although events are not permanently stored on the blockchain, they can still be retrieved for monitoring purposes after being emitted. Here’s an example of how to retrieve events for a specific smart contract using the [@massalabs/massa-web3](https://github.com/massalabs/massa-web3) provider:
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+6. `isError`:
+
+- Filters events by error status.
+- Set to true to retrieve only error events, which can help identify failed operations, or false to retrieve only successful events.
+
+## Event poller
+The [@massalabs/massa-web3](https://github.com/massalabs/massa-web3) library includes an event poller, which facilitates continuous monitoring of smart contract events by polling for updates at regular intervals. This feature is particularly useful for applications that require real-time event tracking or need to respond dynamically to contract activities.
+
+### How the Event Poller Works
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:98]:
```suggestion
### How the event poller works
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+2. **User-Defined Events**:
+
+    - In addition to system-generated error events, developers can define custom events to log specific actions or important milestones within the contract.
+    - User-defined events are typically used to provide visibility into contract behavior, such as when a transfer occurs or when a token is minted.
+
+## Important Considerations for Events
+In Massa, events are not stored on the blockchain ledger, meaning they are ephemeral. Therefore, events should not be used as a source of persistent data. Instead, they are suitable for temporary tasks such as monitoring, logging, or providing real-time feedback.
+For usecases that need data persistency, storage should be used.
+
+## Emitting a Custom Event
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:28]:
```suggestion
## Emitting a custom event
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+6. `isError`:
+
+- Filters events by error status.
+- Set to true to retrieve only error events, which can help identify failed operations, or false to retrieve only successful events.
+
+## Event poller
+The [@massalabs/massa-web3](https://github.com/massalabs/massa-web3) library includes an event poller, which facilitates continuous monitoring of smart contract events by polling for updates at regular intervals. This feature is particularly useful for applications that require real-time event tracking or need to respond dynamically to contract activities.
+
+### How the Event Poller Works
+The event poller retrieves events from the Massa blockchain based on specified filters and invokes callback functions for handling incoming data and errors. You can configure the poller with various parameters, such as the smart contract address and the polling interval, to tailor its behavior to your needs.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:99]:
```suggestion
The event poller retrieves events from the Massa blockchain using specified filters and invokes callback functions for handling incoming data and errors. You can configure the poller with various parameters, such as the smart contract address and the polling interval, to tailor its behavior to your needs.
```

File: docs/build/smart-contract/basic-concepts/events.mdx [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+### Example: Setting Up an Event Poller
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/events.mdx:101]:
```suggestion
### Example: Setting up an event poller
```

File: docs/build/smart-contract/basic-concepts/serializable.mdx [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+}
+```
+
+### Explanation of the User Class
+1. Constructor:
+
+    - The constructor initializes User instances with a name and an age. - Default values are provided to allow for easy instantiation.
+
+2. serialize Method:
+    - This method uses Args to convert the name and age properties into a byte array. It returns a serialized representation of the User object, which can then be stored in the contract's storage.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/serializable.mdx:36]:
```suggestion
    - This method uses `Args` to convert the `name` and `age` properties into a byte array. It returns a serialized representation of the User object, which can then be stored in the contract's storage.
```

File: docs/build/smart-contract/basic-concepts/serializable.mdx [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+
+3. deserialize Method:
+    - The deserialize method takes in a byte array and an offset, reconstructing the User object by extracting the name and age values. This enables the User instance to be rehydrated from storage.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/serializable.mdx:39]:
```suggestion
    - The deserialize method takes in a byte array and an offset, reconstructing the User object by extracting the name and age values. This allows the User instance to be reconstructed from storage.
```

File: docs/build/smart-contract/basic-concepts/storage.mdx [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Retrieve a value
+export function getUser(username: string): string {
+    assert(Storage.has(USER_KEY))
+    return Storage.get(USER_KEY, username);
+}
+```
+
+### Interacting with Storage from Another Address:
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/storage.mdx:48]:
```suggestion
### Interacting with storage from another address:
```

File: docs/build/smart-contract/basic-concepts/storage.mdx [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+    - Retrieves the value associated with the specified key.
+    - If the key doesn’t exist, this will stop the operation execution and revert all previously executed code.
+- `Storage.has(key: T): bool`:
+     - Checks if a value exists for the specified key and returns true if it does, or false if it doesn't.
+- `Storage.del(key: T): void`:
+     - Delete the key-value pair associated with the specified key.
+     - If the key doesn't exist, this will stop the operation execution and revert all previously executed code.
+
+
+### Using Storage:
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/storage.mdx:29]:
```suggestion
### Using storage:
```

File: docs/build/smart-contract/basic-concepts/storage.mdx [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+- The primary methods for interacting with contract storage are `Storage.set`, `Storage.get`, and `Storage.has`.
+- Although storage keys and values are ultimately stored as raw bytes, these methods accept `StaticArray<u8>` `string` and `Args` as arguments, automatically handling serialization and deserialization.
+
+Here's a quick overview of each method, using `T = StaticArray<u8> | string | Args`
+
+- `Storage.set(key: T, value: T): void`:
+    - Stores the specified value under the provided key. If the key already exists, it updates the value.
+    - If new storage are used, it will occurs a cost in MAS. Learn more about storage cost [here](https://docs.massa.net/docs/learn/storage-costs)
+- `Storage.append(key: T, value: T): void`:
+     - Append data at a storage key
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/storage.mdx:17]:
```suggestion
     - Append data to a storage key
```

File: docs/build/smart-contract/basic-concepts/storage.mdx [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+
+### Interacting with Storage from Another Address:
+
+In some cases, you may need to read from or write to the storage of another smart contract or address. For this purpose, Massa provides the Storage.setOf, Storage.appendOf, Storage.getOf, and Storage.hasOf methods, which include an additional parameter for specifying the target address.
+
+Here's a quick overview of each method, using `T = StaticArray<u8> | string | Args`
+
+- `Storage.setOf(address: Address, key: T, value: T): void`:
+    - Stores a value in the storage of a specified address under the given key.
+    - Writing data to another contract storage require the caller to have writte access on the given contract. It happen only when deploying the contract from another one.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/storage.mdx:56]:
```suggestion
    - Writing data to another contract storage require the caller to have writte access on the given contract. This occurs only when deploying the contract from another one.
```

File: docs/build/smart-contract/basic-concepts/storage.mdx [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+
+- `Storage.appendOf(address: Address, key: T, value: T): void`:
+    - Append data at a storage key
+    - Writing data to another contract storage require the caller to have writte access on the given contract. It happen only when deploying the contract from another one.
```
Comment (modship) [inline docs/build/smart-contract/basic-concepts/storage.mdx:60]:
```suggestion
    - Writing data to another contract storage require the caller to have writte access on the given contract. This occurs only when deploying the contract from another one.
```

File: docs/build/smart-contract/intro.mdx [hunk 29]
```diff
@@ -0,0 +1,29 @@
+---
+sidebar_position: 1
+---
+
+
+# Introduction
+
+Smart contracts are programs that interact with the blockchain, accessing ledger storage and modifying the current state of the ledger. They enable developers to build decentralized applications (dApps) that enforce contractual terms autonomously, embodying the principle that **code is law**.
```
Comment (modship) [inline docs/build/smart-contract/intro.mdx:8]:
```suggestion
Smart contracts are programs that interact with the blockchain, accessing and modifying the ledger's storage state. They enable developers to build decentralized applications (dApps) that enforce contractual terms autonomously, embodying the principle that **code is law**.
```

File: docs/build/smart-contract/intro.mdx [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+
+## How smartcontract works on Massa?
```
Comment (modship) [inline docs/build/smart-contract/intro.mdx:10]:
```suggestion
## How smart contract works on Massa?
```

File: docs/build/smart-contract/intro.mdx [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+In this framework, a smart contract can be viewed as a stored procedure that can modify the ledger, interact with other smart contracts, and be invoked by end users.
+
+With **asynchronous execution**, smart contracts on Massa can automatically activate at a specified time or when certain conditions in the ledger are met, such as a specific data value within the contract.
+
+## Execution constraints of a Smart Contracts
+
+Smart contracts are executed directly by Massa nodes, with fees in place to prevent spam and network congestion:
+
+- The processing unit, or [gas](https://docs.massa.net/docs/learn/gas), measures the complexity of contract execution. Each block has a gas limit, which constrains the amount of computation that can occur.
+- According to the amount of gas needed to execute the smart contract, the caller must pay a fee in MAS for the operation to be included in a block by the Massa validators.
```
Comment (modship) [inline docs/build/smart-contract/intro.mdx:25]:
```suggestion
- The caller must pay a fee in MAS based on the amount of gas required to execute the smart contract for the operation to be included in a block by the Massa validators.
```

File: docs/build/smart-contract/intro.mdx [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+
+## Execution constraints of a Smart Contracts
+
+Smart contracts are executed directly by Massa nodes, with fees in place to prevent spam and network congestion:
+
+- The processing unit, or [gas](https://docs.massa.net/docs/learn/gas), measures the complexity of contract execution. Each block has a gas limit, which constrains the amount of computation that can occur.
+- According to the amount of gas needed to execute the smart contract, the caller must pay a fee in MAS for the operation to be included in a block by the Massa validators. 
+- If a smart contract requires [storage](https://docs.massa.net/docs/learn/storage-costs), it incurs an additional cost.
+
+Finally, smart contracts on Massa are deterministic. Once deployed, their code and interface are publicly available and unchangeable.
```
Comment (modship) [inline docs/build/smart-contract/intro.mdx:28]:
```suggestion
Finally, smart contracts on Massa are deterministic, meaning that their outcomes are predictable and consistent. Once deployed, their code and interface are publicly available and unchangeable.
```

File: docs/build/smart-contract/prerequisites.mdx [hunk 33]
```diff
@@ -5,6 +5,7 @@ sidebar_label: Prerequisites
 
 # Prerequisites
 
+Assemblyscript use the npm registry format to manage dependencies and project entrypoint is the `packages.json`.
```
Comment (modship) [inline docs/build/smart-contract/prerequisites.mdx:8]:
```suggestion
AssemblyScript use the npm registry format to manage dependencies and project entrypoint is the `packages.json`.
```

**Review by modship — COMMENTED** [review]
:rocket:

**Review by modship — APPROVED** [review]

**Review by Ben-Rey — APPROVED** [review]

**Pull request merged**